### ISSUE TITLE
IOException: Too many connections per host <#>

### ISSUE THREAD

**Issue by yoeduardoj** [AsyncHttpClient/async-http-client#850]
After we set .setMaxConnectionsPerHost(64), our server seemed to happily work. We can pound it with traffic and see very few issues with connections since its so efficient at pooling connections that are in  good condition.

Lib: async-http-client 1.9.15
java version "1.7.0_67"

After a while, however (about 24 hours), we start getting the above exception coming from the ChannelManager.

Looking at the NettyResponseListener code, I noticed something odd. 

https://github.com/AsyncHttpClient/async-http-client/blob/b85d5b3505d9f6e80d278fef88876f6546e73079/providers/netty4/src/main/java/org/asynchttpclient/providers/netty4/request/NettyRequestSender.java

In ```NettyRequestSender.sendRequestWithNewChannel()```, there's this bit:

```java
        boolean channelPreempted = false;
        String partition = null;

        try {            // Do not throw an exception when we need an extra connection for a
            // redirect.
            if (!reclaimCache) {

                // only compute when maxConnectionPerHost is enabled
                // FIXME clean up
                if (config.getMaxConnectionsPerHost() > 0)
                    partition = future.getPartitionId();

                channelManager.preemptChannel(partition);
            }

            if (asyncHandler instanceof AsyncHandlerExtensions)
                AsyncHandlerExtensions.class.cast(asyncHandler).onOpenConnection();

            ChannelFuture channelFuture = connect(request, uri, proxy, useProxy, bootstrap, asyncHandler);
            channelFuture.addListener(new NettyConnectListener<T>(future, this, channelManager, channelPreempted, partition));

        } catch (Throwable t) {
            if (channelPreempted)
                channelManager.abortChannelPreemption(partition);

            abort(null, future, t.getCause() == null ? t : t.getCause());
        }
```

If you notice, channelPreempted never gets written to. Isn't channelPreempted = true missing from the block where the channel is preempted?

Shouldn't it be:

```java
        boolean channelPreempted = false;
        String partition = null;

        try {            // Do not throw an exception when we need an extra connection for a
            // redirect.
            if (!reclaimCache) {

                // only compute when maxConnectionPerHost is enabled
                // FIXME clean up
                if (config.getMaxConnectionsPerHost() > 0)
                    partition = future.getPartitionId();

                channelManager.preemptChannel(partition);
                channelPreempted = true;
            }

            if (asyncHandler instanceof AsyncHandlerExtensions)
                AsyncHandlerExtensions.class.cast(asyncHandler).onOpenConnection();

            ChannelFuture channelFuture = connect(request, uri, proxy, useProxy, bootstrap, asyncHandler);
            channelFuture.addListener(new NettyConnectListener<T>(future, this, channelManager, channelPreempted, partition));

        } catch (Throwable t) {
            if (channelPreempted)
                channelManager.abortChannelPreemption(partition);

            abort(null, future, t.getCause() == null ? t : t.getCause());
        }
```

The same class for netty3 has the correct code:

https://github.com/AsyncHttpClient/async-http-client/blob/b85d5b3505d9f6e80d278fef88876f6546e73079/providers/netty3/src/main/java/org/asynchttpclient/providers/netty3/request/NettyRequestSender.java

**Comment by yoeduardoj** [comment]
Actually, this shouldn't be an issue with 1.9.x. (but in your master branch I def. think it is a bug).

With 1.9.x, maybe #844 is related to our error.

**Comment by slandelle** [comment]
I fixed the missing channelPreempted issue on master. Thanks for reporting.

**Comment by slandelle** [comment]
> After a while, however (about 24 hours), we start getting the above exception coming from the ChannelManager.

Which above exception? I think you forgot to paste.

**Comment by slandelle** [comment]
`setMaxConnectionsPerHost` only caps the number of connections that can be open to a given host. There's no built-in queuing mechanism for requests that might need a connection while there's none available.

So, except if you find found out that there's a semaphore leak, it could be a normal behavior under a sudden load rise.

**Comment by yoeduardoj** [comment]
I get that, but the exception is reported when netstat clearly shows the connections are not there.

**Comment by slandelle** [comment]
Could be a semaphore leak then.
A reproducer would definitively help...

**Comment by jonas-ren** [comment]
Seeing the same problem with Netty 3 as well. Things start fine and after a while every connections is reporting: IOException: Too many connections per host <#>.

**Comment by slandelle** [comment]
Answer is still the same: would it be possible for you to you provide a reproducer?

**Comment by jonas-ren** [comment]
This is quite easy to reproduce. You can just set maxConnectionsPerHost to N, and then continuously send HTTP requests from one AsyncHttpClient instance to a single IP address. After N succeeding requests, all the following requests will report "IOException: Too many connections per host <N>".

As mentioned by yoeduardoj, it seems like it's not account for connection releases at all. This can be observed with the latest version 1.9.27.

**Comment by nkrode** [comment]
I can confirm this is the same issue I'm experiencing, If I set setMaxConnections and setMaxConnectionsPerHost and make multiple requests to the same Host, "Too many connections" error is reported. Here's the config I'm using:

 AsyncHttpClientConfig cf = new AsyncHttpClientConfig.Builder()
                .setAllowPoolingConnections(true)
                .setAllowPoolingSslConnections(true)
                .setMaxConnections(10)
                .setMaxConnectionsPerHost(10)
                .build();

If I remove the maxConnections limit then it works fine until I run into the OS max file descriptors limit.

**Comment by slandelle** [comment]
@jonas-ren @nkrode Let's make one thing clear: "IOException: Too many connections per host" is the expected behavior if you try to open too many concurrent connections to the same host.

It's only a security, there's no queueing mechanism to would take care of the overflow. And even if there was one (I'm not convinced), what would the behavior be once the queue is full? Drop? And how would the caller be notified?

Then, if you're telling me, like @yoeduardoj, that you shouldn't have exceeded the limit but still got the exception, then there's indeed an issue somewhere. And a reproducer would help = some client code that makes sure it doesn't overflow, and still get the exception.

**Comment by nkrode** [comment]
@slandelle Ok maybe I'm misunderstanding the connection pooling here, just to double confirm, if I set my configuration as maxConnectionsPerHost = 10 and I Issues 11 concurrent requests, then are you saying that the first 10 connections will go through and the 11th connection will get the too many connections per host error?

If that's the case, then what is the intent of pooling connections? Is it to reuse the already established http connection? I'm assuming that's provided http keep alive is supported by the server.

**Comment by slandelle** [comment]
> are you saying that the first 10 connections will go through and the 11th connection will get the too many connections per host error?

Yes

> If that's the case, then what is the intent of pooling connections?

Reuse open keep-alive connections.

Just think of a standard database connection pool where the borrow timeout is 0 (ie, once you reach the limit, you immediately get an exception instead of blocking until a connection becomes available, or you get a timeout).

**Comment by nkrode** [comment]
cool, makes sense :+1:

**Comment by jonas-ren** [comment]
@slandelle Is this the expected behavior if connection pooling is disabled? I'm seeing this with allowPoolingConnections being false, and the HTTP connection header is always "close". The connection should be closed and we should not hit the maxConnectionPerHost limit, no?

This is the HTTP header I got directly from tcpdump:
GET / HTTP/1.0
Connection: close
User-Agent: XXXX
Accept: */*
Accept-Encoding: *
Host: www.example.com

**Comment by slandelle** [comment]
@jonas-ren OK, connection pool config limits are still honored when connection pooling is disabled. Fixing.

**Comment by slandelle** [comment]
Fixed: https://github.com/AsyncHttpClient/async-http-client/issues/918

**Comment by jonas-ren** [comment]
@slandelle Thanks for fixing this! But actually our use case is that we do enable connection pooling but only want to pool the connections with keep-alive connection headers. If the connection is closed, it should not count towards maxConnectionsPerHost. Is it gonna work as expected with your fix?

**Comment by slandelle** [comment]
OK, now I remember properly (I didn't introduce this feature, and don't use it).
Those limits have a more general purpose than the sole pooling of keep-alive connections: they rule the number of open connections. You can't know if a connection can be pooled before you receive the response.

Reverting last commit.

**Comment by jonas-ren** [comment]
Well, I found where the problem is. We use a custom filter to track total connections but use AHC to limit connections per host. However, the channel manager doesn't override the remove method for CleanupChannelGroup if maxTotalConnections is not enabled but maxConnectionsPerHost is enabled. 

See https://github.com/AsyncHttpClient/async-http-client/blob/master/providers/netty3/src/main/java/org/asynchttpclient/netty/channel/ChannelManager.java#L132

**Comment by slandelle** [comment]
@jonas-ren Good catch, thanks!

**Comment by slandelle** [comment]
Fixed, thanks: https://github.com/AsyncHttpClient/async-http-client/issues/920

**Comment by slandelle** [comment]
Assuming last issue reported by @yoeduardoj is the one identified by @jonas-ren and closing.
Please ping otherwise.

**Comment by jonas-ren** [comment]
thanks for fixing!

**Comment by slandelle** [comment]
@jonas-ren All credits go to you :smile:

**Comment by richiesgr** [comment]
Hi sorry to continue this here
but the setKeepAlive method has disappeared from AsyncHttpClientConfig in 1.9.xxx
How I'm supposed to set this ?
Thanks

**Comment by slandelle** [comment]
`setMaxConnectionsPerHost` is a security. AHC doesn't have (and probably won't) have any buffering mechanism if you're trying to exceed this limit, for example if you try to send too many concurrent requests.

**Comment by richiesgr** [comment]
I understand this. 
So yes I make concurrent request in my case 5  but in this case is there a way to close or free connection for the next 5 if not how to use the client need to recreate a client each time we use it ?

**Comment by akstheking** [comment]
I am using version 1.9.31 of AHC. I see changes of fix #920 in the source of the library but I still the errors coming : ** Error Too many connections per host 100000**.
This is what my builder looks like :
```
new AsyncHttpClientConfig.Builder()
			.setCompressionEnforced(false)
			.setFollowRedirect(true)
			.setAllowPoolingConnections(true)
			.setConnectTimeout(1200)
			.setPooledConnectionIdleTimeout(1200)
			.setRequestTimeout(1200)
			.setMaxConnectionsPerHost(100000)
			.setMaxConnections(500000)
			.setReadTimeout(1500)
			.setConnectionTTL(1200+100)
					.setExecutorService(new ThreadPoolExecutor(50, 500,
							1200 + 100, TimeUnit.MILLISECONDS,
					new LinkedBlockingQueue<Runnable>()))
			.addIOExceptionFilter(new ResumableIOExceptionFilter());
```
